### PR TITLE
1258: add missing fields in 'serialize' methods

### LINKED ISSUE #1258: fields omitted during serialization
**Describe the bug**
Multiple fields are not serialized during either footprinting or regular serialization. Add missing fields to `serialize` methods.
Note: this should be safe for all of footprinting, caution is required for serializers used in other modes.

**To Reproduce**
See [serialization sanitizer run over `vt`](https://github.com/DARMA-tasking/serialization-sanitizer/actions?query=workflow%3A%22PR+tests+%28vt%2C+clang-5.0%2C+ubuntu%29%22).

### PR DESCRIPTION
**Pull request by cz4rs**
fixes #1258

### CODE REVIEW

**Comment by cz4rs** [comment]
This is based on the experimental reports (after reducing the number of false positives).
- add a bunch of fields completely missed for footprinting purposes
- change `countBytes` to `|` operator - possible after https://github.com/DARMA-tasking/checkpoint/pull/157 was merged (footprinting is available for all non-serializable types `T` using `sizeof(T)` behind the scenes)
- add missing fields for regular serialization in tests and some simple cases


cases still reported:

- [ ] `vt::messaging::MsgSharedPtr::impl_`
- [ ] `vt::messaging::MsgSharedPtr::ptr_ `
- [ ] `t::pipe::callback::cbunion::GeneralCallback::active_`
- [ ] `vt::pipe::callback::cbunion::GeneralCallback::u_ `
- [ ] `vt::pipe::signal::Signal::data_ptr_`
- [ ] `vt::rdma::BaseTypedHandle::actions_ `
- [ ] `vt::rdma::BaseTypedHandle::lock_ `
- [ ] `vt::rdma::BaseTypedHandle::user_buffer_ `
- [ ] `vt::rdma::Manager::holder_footprint_ ` - custom serialization for static holder
- [ ] `vt::term::interval::IntegralSetBase::hint_ ` - custom unpacking
- [ ] `vt::termination::graph::EpochGraph::successors_ ` - custom unpacking
- [ ] `vt::tests::unit::MyCol::op6_msgs_ ` - custom behavior
- [ ] `vt::trace::Trace::log_file_` - definition unavailable, `error: invalid application of ‘sizeof’ to incomplete type ‘vt::trace::vt_gzFile’`
- [ ] `vt::tutorial::ParticleMsg::a ` - custom serialization of raw pointer
- [ ] `vt::vrt::collection::MigrateMsg::owns_elm_ `

File: src/vt/pipe/callback/cb_union/cb_raw.h [hunk 1]
```diff
@@ -255,6 +255,9 @@ struct GeneralCallback {
       vtAssert(0, "Should be unreachable");
       break;
     }
+
+    s.skip(active_);
```
Comment (PhilMiller) [inline src/vt/pipe/callback/cb_union/cb_raw.h:259]: Stylistically, I'd prefer having `skip` calls directly alongside the block of code that takes care of the variable in question.

Comment (cz4rs) [reply src/vt/pipe/callback/cb_union/cb_raw.h]: moved close to the `switch` case

File: src/vt/pipe/callback/cb_union/cb_raw.h [hunk 2]
```diff
@@ -222,7 +222,10 @@ struct GeneralCallback {
     using EnumDataType = typename std::underlying_type<CallbackEnum>::type;
     EnumDataType val = static_cast<EnumDataType>(active_);
     s | val;
+    s.skip(active_);
+
     active_ = static_cast<CallbackEnum>(val);
+    s.skip(u_); // only serialize actual content of the union
```
Comment (PhilMiller) [inline src/vt/pipe/callback/cb_union/cb_raw.h:228]: On second look, I realize that specifically in the footprinting case, serializing just the active member hypothetically undercounts the size of the callback itself, even as the payload of the callback is likely to be substantially larger.

A rigorous solution would be something like `s.addBytes(sizeof(u_) - sizeof(u_.active_member_))` I think - i.e. round up the leftover and padding bytes and add that.

As with everything else, the case of concern is if we happen to have many instances of a type accumulating unexpectedly, and blowing up memory, and that's the one we really want footprinting to illuminate.

File: src/vt/pipe/callback/cb_union/cb_raw.h [hunk 3]
```diff
@@ -219,37 +219,43 @@ struct GeneralCallback {
...
   void serialize(SerializerT& s) {
-    using EnumDataType = typename std::underlying_type<CallbackEnum>::type;
-    EnumDataType val = static_cast<EnumDataType>(active_);
-    s | val;
-    active_ = static_cast<CallbackEnum>(val);
+    s | active_;
+
+    // serialize actual content of the union and account for leftovers and
+    // padding bytes when footprinting
+    auto ser = [&s, this](auto& cb){
```
Comment (cz4rs) [inline src/vt/pipe/callback/cb_union/cb_raw.h:226]: `nvcc` would complain without explicit capture of `this`:
```
/vt/src/vt/pipe/callback/cb_union/cb_raw.h:228:26: error: base operand of '->' is not a pointer
       s.addBytes(sizeof(this->u_) - sizeof(cb));
                          ^~
```

**Review by lifflander — APPROVED** [review]

File: src/vt/pipe/callback/cb_union/cb_raw.h (hunk 2 above)
Comment (cz4rs) [reply src/vt/pipe/callback/cb_union/cb_raw.h]: :+1: for noticing this, fixed it along with `NoCB` case

**Pull request merged**